pkg/sources/source_manager.go

```diff
@@ -0,0 +1,168 @@
+package sources
+
+import (
+	"fmt"
+	"runtime"
+	"sync"
+	"sync/atomic"
+
+	"github.com/trufflesecurity/trufflehog/v3/pkg/common"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/context"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/pb/sourcespb"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+// handle uniquely identifies a Source given to the manager to manage. If the
+// SourceManager is connected to the API, it will be equivalent to the unique
+// source ID, otherwise it behaves as a counter.
+type handle int64
+
+type SourceManager struct {
+	api apiClient
+	// Map of handle to source information.
+	handles     map[handle]sourceTemplate
+	handlesLock sync.Mutex
+	// Counters for assigning handle and job IDs for headless managers.
+	headlessHandleCounter int64
+	headlessJobCounter    int64
+	// Pool limiting the amount of concurrent sources running.
+	pool errgroup.Group
+	// Downstream chunks channel to be scanned.
+	outputChunks chan *Chunk
+}
+
+// sourceTemplate is an aggregate struct to hold the Source and its name.
+type sourceTemplate struct {
+	name   string
+	source Source
+}
+
+// apiClient is an interface for optionally communicating with an external API.
+type apiClient interface {
+	// RegisterSource lets the API know we have a source and it returns a unique source ID for it.
+	RegisterSource(ctx context.Context, name string, kind sourcespb.SourceType) (int64, error)
+	// GetJobID queries the API for an existing job or new job ID.
+	GetJobID(ctx context.Context, id int64) (int64, error)
+}
+
+// WithAPI adds an API client to the manager for tracking jobs and progress.
+func WithAPI(api apiClient) func(*SourceManager) {
+	return func(man *SourceManager) { man.api = api }
+}
+
+// WithConcurrency limits the concurrent number of sources a manager can run.
+func WithConcurrency(concurrency int) func(*SourceManager) {
+	return func(man *SourceManager) { man.pool.SetLimit(concurrency) }
+}
+
+// NewManager creates a new manager with the provided options.
+func NewManager(outputChunks chan *Chunk, opts ...func(*SourceManager)) *SourceManager {
+	man := SourceManager{
+		handles:      make(map[handle]sourceTemplate),
+		outputChunks: outputChunks,
+	}
+	for _, opt := range opts {
+		opt(&man)
+	}
+	return &man
+}
+
+// Manage informs the SourceManager to track and manage a Source.
+func (s *SourceManager) Manage(ctx context.Context, name string, source Source) (handle, error) {
+	id, err := s.registerSource(ctx, name, source)
+	if err != nil {
+		return 0, err
+	}
+	s.handlesLock.Lock()
+	defer s.handlesLock.Unlock()
+	if _, ok := s.handles[id]; ok {
+		// TODO: smartly handle this?
+		return 0, fmt.Errorf("generated handle ID '%d' already in use locally", id)
```

Comment:
I'm wondering if you could avoid having to worry about this weird case by establishing a facade that creates some sort of virtual API for headless managers.

Reply:
Ah do you mean like making the `api` required and moving the `headlessJobCounter` etc. into an implementation of that?

I don't see how that would remove this edge case though so maybe I'm misunderstanding.